---
title: >
  Transcriptome Data Analysis - hands-on session - IN A GLANCE
subtitle: >
  MSE - Module Genetic Epidemiology
  <p align="center">
  </p>
  <a href="https://www.unimedizin-mainz.de/imbei/startseite/mse/"><img src="images/mse_logo.jpg" alt="" height="100"/></a>
author:
- name: <a href="https://federicomarini.github.io">Federico Marini (marinif@uni-mainz.de)</a><br><a href="https://www.unimedizin-mainz.de/imbei/">IMBEI, University Medical Center Mainz</a><br><a href="https://twitter.com/FedeBioinfo">`r icons::fontawesome('twitter')` `@FedeBioinfo`</a>
- name: Alicia Schulze (alpoplaw@uni-mainz.de)<br><a href="https://www.unimedizin-mainz.de/imbei/">IMBEI, University Medical Center Mainz</a><br>
date: "2024/07/03-04"
output: 
  bookdown::html_document2:
    toc: true
    toc_float: true
    theme: cosmo
    code_folding: show
    code_download: true
editor_options: 
  chunk_output_type: inline
bibliography: references.bib
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#",
  error = FALSE,
  warning = FALSE,
  message = FALSE,
  fig.width = 10, 
  fig.height = 8
)
```

<!-- <script type="text/javascript" -->
<!--   src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"> -->
<!-- </script> -->


```{r style, echo=FALSE, message=FALSE, warning=FALSE, results="asis", eval = TRUE}
suppressPackageStartupMessages({
  library(BiocStyle)
  library(knitr)
  library(rmarkdown)
})
options(width = 80)
```

# Introduction

This tutorial is intended to be used "at a glance":

* It does not contain too much text
* It contains almost only the essential commands
* It does run through the whole workflow
* It is the perfect trampoline to learn more in detail - as you can do in the full `transcriptomics_practical` document

You can consider this as a very much squeezed (yet well squeezed) and complete (yup!) version of the whole (the real, whole) analysis workflow when dealing with RNA-seq data.

Ready? Let's go!

# The whole game, in a nutshell

## The dataset: `macrophage`

```{r thedataset}
library("macrophage")
dir <- system.file("extdata", package = "macrophage")
coldata <- read.csv(file.path(dir, "coldata.csv"))[, c(1, 2, 3, 5)]
## how many samples do we have?
## what are the important experimental factors?
coldata

coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
## these are the files output from salmon, containing all info on the quantifications
coldata$files

suppressPackageStartupMessages({
  library("tximeta")
  library("DESeq2")
  library("org.Hs.eg.db")
  library("SummarizedExperiment")
})

## at the transcript level...
st <- tximeta(coldata = coldata, type = "salmon", dropInfReps = TRUE)
head(assay(st, "counts"), 3)
rowData(st)
colData(st)

## at the gene level...
sg <- tximeta::summarizeToGene(st)
sg
## can you guess what happened?
sg <- tximeta::addIds(sg, "SYMBOL", gene = TRUE)
sg
head(rowData(sg))

## add gene symbols
sg <- tximeta::addIds(sg, "SYMBOL", gene = TRUE)
sg
head(rowData(sg))
colData(sg)

table(colData(sg)$condition_name)
table(colData(sg)$line_id)
## what is our aim? Find DE genes by which design specification?

colData(sg)$condition_name <- factor(colData(sg)$condition_name)
colData(sg)$condition_name <- relevel(colData(sg)$condition_name, ref = "naive")
colData(sg)$condition_name

## here's the main container for all infos!
dds <- DESeqDataSet(sg, design = ~ line_id + condition_name)

dds
## all set!
```

## Exploration, modeling, testing

```{r emt}
nrow(dds)
table(rowSums(assay(dds, "counts")) == 0)

keep <- rowSums(counts(dds)) > 1
dds <- dds[keep, ]
dim(dds)

## exploratory analysis and visualization
vsd <- DESeq2::vst(dds)
DESeq2::plotPCA(vsd, intgroup = "condition_name")
pcaExplorer::pcaplot(vsd, intgroup = "condition_name")

## differential expression analysis boils down to...
dds <- DESeq2::DESeq(dds)

DESeq2::plotDispEsts(dds)

## Default - SL1344 vs naive
res <- DESeq2::results(dds)
head(res)

## We'll instead focus on IFNgamma vs naive
res <- DESeq2::results(dds, contrast = c("condition_name", "IFNg", "naive"))
head(res)
## what do these columns mean?

summary(res)
hist(res$pvalue)
## Remove the genes that were filtered out in the independent filtering
hist(res$pvalue[!is.na(res$padj)])

## can we be more strict?

res.05 <- results(dds, alpha = 0.05, 
                  contrast = c("condition_name", "IFNg", "naive"))
table(res.05$padj < 0.05)
## lower the false discovery rate threshold


resLFC1 <- results(dds, lfcThreshold = 1, 
                   contrast = c("condition_name", "IFNg", "naive"))
summary(resLFC1)
table(resLFC1$padj < 0.1)
## raise the log2 fold change threshold from 0 using the `lfcThreshold` argument of *results*

## checking individual genes
plotCounts(dds, gene = "ENSG00000126561.16", intgroup = "condition_name", 
           normalized = TRUE, transform = FALSE)

GeneTonic::gene_plot(dds, gene = "ENSG00000126561.16", intgroup = "condition_name", 
           normalized = TRUE, transform = FALSE)

## checking an overview on all genes
DESeq2::plotMA(res, ylim = c(-5, 5))

library("apeglm")
DESeq2::resultsNames(dds)
res_ape <- DESeq2::lfcShrink(dds, coef = "condition_name_IFNg_vs_naive", type = "apeglm")
DESeq2::plotMA(res_ape, ylim = c(-5, 5))
## what happened here?

library("pheatmap")
stopifnot(rownames(vsd) == rownames(res))

## looking at the top 30 DE genes
mat <- assay(vsd)
rownames(mat) <- rowData(vsd)$SYMBOL
mat <- mat[head(order(res$padj), 30), ]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(vsd)[, c("condition_name"), drop = FALSE])
pheatmap(mat, annotation_col = df)

## looking at the most variable ones
mat <- assay(vsd)
rownames(mat) <- rowData(vsd)$SYMBOL
topVarGenes <- head(order(rowVars(mat), decreasing = TRUE), 30)
mat <- mat[topVarGenes, ]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(vsd)[, c("condition_name"), drop = FALSE])
pheatmap(mat, annotation_col = df)
## is there something in common here?
```

## Making sense out of so many genes

```{r enrichgt}
library("GeneTonic")

res$symbol <- rowData(dds)$SYMBOL

de_symbols_IFNg_vs_naive <- deseqresult2df(res, FDR = 0.05)$symbol
bg_ids <- rowData(dds)$SYMBOL[rowSums(counts(dds)) > 0]

library("topGO")
topgo_DE_macrophage_IFNg_vs_naive <- pcaExplorer::topGOtable(
    DEgenes = de_symbols_IFNg_vs_naive,
    BGgenes = bg_ids,
    ontology = "BP",
    mapping = "org.Hs.eg.db",
    geneID = "symbol",
    topTablerows = 500
)

library("clusterProfiler")
clupro_DE_macrophage_IFNg_vs_naive <- clusterProfiler::enrichGO(
    gene = de_symbols_IFNg_vs_naive,
    universe = bg_ids,
    keyType = "SYMBOL",
    OrgDb = org.Hs.eg.db,
    ont = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff = 0.01,
    qvalueCutoff = 0.05,
    readable = FALSE
)

DT::datatable(topgo_DE_macrophage_IFNg_vs_naive[1:10,])
DT::datatable(as.data.frame(clupro_DE_macrophage_IFNg_vs_naive)[1:10,])
```

## Ooooh so many tables, how to avoid madness?

```{r exportthings}
## can I export these tables to text/excel?
stopifnot(all(rownames(res) == rownames(dds)))
res$symbol <- rowData(dds)$SYMBOL

resOrdered <- res[order(res$padj), ]
head(resOrdered)

resOrderedDF <- as.data.frame(resOrdered)[seq_len(100), ]
write.table(cbind(id = rownames(resOrderedDF), resOrderedDF), 
            file = "myderesults.txt", quote = FALSE, sep = "\t",
            row.names = FALSE)

## can we streamline things here?
res_enrich_topGO <- shake_topGOtableResult(topgo_DE_macrophage_IFNg_vs_naive)
res_enrich_clupro <- shake_enrichResult(clupro_DE_macrophage_IFNg_vs_naive)

gtl_macrophage <- GeneTonicList(
    dds = dds,
    res_de = res,
    res_enrich = res_enrich_clupro,
    annotation_obj = data.frame(
        gene_id = rowData(dds)$gene_id,
        gene_name = rowData(dds)$SYMBOL
    )
)

## saving this
# saveRDS(gtl_macrophage, "gtl_macrophage.RDS")
## reloading this
# gtl_macrophage <- readRDS("gtl_macrophage.RDS")
## and that is it!
# GeneTonic(gtl = gtl_macrophage)
## or if expecting to upload at runtime... (e.g. used as a server-like app)
# GeneTonic()
```


## Some quiz questions as we proceed{-}

* Can you start your DE analysis with normalized values such as TPMs?
* Where can I store additional information about the samples?
* Where can I store additional information about the features? (What can your features be?)
* Can you think of a way to read in the required information if you start (AAAAAAAH) from Excel files?
* How important is the definition of the design?
* How do I quantify the effect size? "In which direction" is this to be interpreted?
* Is it possible to change the reference in the comparison?
* If you have multiple experimental factors, how should you proceed? Think of 2 condition-2 tissues/cell types
* Removing unexpressed features might influence what you do when integrating different analyses. How could we proceed in such cases?
* What if I download publicly available data and some genes are missing? What can be some of the possible reasons for this? 
* What if I do not see a clear separation in my PCA plot? Is it still ok to proceed?
* What should I do if I detect/am aware of a batch effect?
* Raw data vs normalized data vs transformed data: Which one should I use in "all the cases" I could encounter?
* "I have seen people using tSNE/UMAP in single cell data, why shouldn't we do the same here?"
* A feature being called DE: "Why is my expected shortlisted gene not showing up?" What possible factors can play a role in the feature being above or below the significance threshold?
* Shall I subset my DE results to the genes only detected as DE? Why?
* The results between DESeq2 and edgeR might differ. Which one is "correct"? How can you better appreciate the existing commonalities and differences in the DE analysis?
* The thought above is valid also for the process of integrating different DE analyses (i.e. using the same method, but "comparing different groups"). What can you think of to represent these results?
* Is visualization of data and results *really that important*?
* Volcano plot or MA plot, what is "better"? Think of the information every plot type displays
* Can I always plot expression values for a single gene, or for a set of genes (i.e. independently of their DE status)?
* What can I do if I want to know more on one "novel" gene, possibly relevant with my experiment?
* Genes or geneset: What should I use as a key to interpretation?
* What is the main advantage of using standardized objects vs a bunch of tables?
* "Ok, I would still like your Bioinformatics group on board for my project. What should I expect to happen?"

# Session information {-}

```{r sessioninfo}
sessionInfo()
```

